core/src/main/scala/org/apache/spark/metrics/MetricsSystem.scala

```diff
@@ -166,9 +172,12 @@ private[spark] class MetricsSystem private (
   }
 
   def removeSource(source: Source) {
-    sources -= source
```

Comment:
I think `sources -= source is` is already looks good.
The method `-=` of `mutable.ArrayBuffer` as follows:
```
  def -= (x: A): this.type = {
    val i = indexOf(x)
    if (i != -1) remove(i)
    this
  }
```
I think the original code works good.

Reply:
but the original code will remove the source's metrics even if it is not registered.

Reply:
I don't understand your meaning.If source not registered,The method `-=` of `mutable.ArrayBuffer` do nothing.

Reply:
```java
  def removeSource(source: Source) {
    sources -= source
    val regName = buildRegistryName(source)
    registry.removeMatching((name: String, _: Metric) => name.startsWith(regName))
  }
```
Even if  the `source` is not in `sources`
the original code will execute the code after `sources -= source` which will remove the metrics of the source.

Reply:
OK. I get it.
This change looks good to me.

Reply:
What metrics would it remove erroneously, if the source were never registered?